File: app/disclosure/requests.py
# app/disclosure/requests.py
"""The request lifecycle: create, list, decide, revoke (directive §5, §6, §15). Every write here
goes through exactly one of these six functions; the two route files (Tasks 5, 6) are thin HTTP
wrappers that resolve a Principal, call one of these, and render the result or the Refusal it
raised. Reuses `app.api.seller_listings.Refusal` rather than declaring a second exception type, so
every route's EXISTING `_refused`/`_error` rendering already knows how to answer one (Global
Constraint: no inline HTTPException; deviation from the plan's literal `from app.api.listings
import Refusal` recorded in task-04-report.md -- `Refusal` is defined at
`app/api/seller_listings.py:201` and `app.api.listings` neither defines nor re-exports it).

Revocation (directive §15) is honest about its own limit and this module never implies
otherwise: `revoke` prevents FUTURE authorized retrieval by moving the row out of the status
`app.disclosure.access.authorized_capabilities` looks for (`'APPROVED'`) -- once revoked, that
function's single query can never again match this row, so no later call through the one
authorization boundary can return a capability for it (`tests/disclosure/test_requests.py`'s
`test_after_revoke_access_authorized_capabilities_is_empty_for_that_buyer` proves exactly this,
against the real function, not against this module's own return value). It cannot, and does not
claim to, recall a file the buyer already downloaded before the seller revoked -- that copy left
this system's control the moment it was served."""
from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from app.api.seller_listings import Refusal
from app.disclosure.levels import CAPABILITIES, REQUESTABLE_LEVELS, covers

_COLUMNS = (
    "id, listing_id, buyer_user_id, seller_user_id, message, status, requested_disclosure_level,"
    " approved_capabilities, requested_at, reviewed_at, reviewed_by, denial_reason, expires_at"
)


def _row(cur: Any) -> dict[str, Any]:
    names = [d[0] for d in cur.description]
    values = cur.fetchone()
    return {name: (str(value) if name.endswith("_id") or name in ("id", "reviewed_by") else value) for name, value in zip(names, values, strict=True)}


def create(conn: Any, *, listing_id: str, buyer_account_id: str, message: str | None,
          requested_disclosure_level: str = "FULL_CONFIDENTIAL") -> dict[str, Any]:
    if requested_disclosure_level not in REQUESTABLE_LEVELS:
        raise Refusal("BAD_LEVEL", f"disclosure_level must be one of {', '.join(sorted(REQUESTABLE_LEVELS))}.", 400)
    with conn.cursor() as cur:
        cur.execute("SELECT seller_id FROM listing WHERE id = %s AND status = 'published'", (listing_id,))
        found = cur.fetchone()
        if found is None:
            raise Refusal("NOT_FOUND", "No such listing.", 404)
        seller_id = found[0]
        if seller_id is None:
            raise Refusal("NO_SELLER", "This listing has no seller account to request access from.", 422)
        if str(seller_id) == str(buyer_account_id):
            raise Refusal("SELF_REQUEST", "You cannot request access to your own listing.", 422)
        try:
            cur.execute(
                f"INSERT INTO request (listing_id, buyer_user_id, seller_user_id, message, requested_disclosure_level)"
                f" VALUES (%s,%s,%s,%s,%s) RETURNING {_COLUMNS}",
                (listing_id, buyer_account_id, seller_id, message, requested_disclosure_level),
            )
        except Exception as exc:  # psycopg2.errors.UniqueViolation — the partial index (migration 096)
            if type(exc).__name__ == "UniqueViolation":
                raise Refusal("ALREADY_REQUESTED", "You already have a pending or approved request for this listing.", 409) from exc
            raise
        return _row(cur)


def get_one(conn: Any, *, request_id: str, buyer_account_id: str) -> dict[str, Any]:
    with conn.cursor() as cur:
        cur.execute(f"SELECT {_COLUMNS} FROM request WHERE id = %s AND buyer_user_id = %s", (request_id, buyer_account_id))
        if cur.fetchone() is None:
            raise Refusal("NOT_FOUND", "No such request.", 404)
        cur.execute(f"SELECT {_COLUMNS} FROM request WHERE id = %s AND buyer_user_id = %s", (request_id, buyer_account_id))
        return _row(cur)


def list_mine(conn: Any, *, buyer_account_id: str) -> list[dict[str, Any]]:
    with conn.cursor() as cur:
        cur.execute(f"SELECT {_COLUMNS} FROM request WHERE buyer_user_id = %s ORDER BY requested_at DESC", (buyer_account_id,))
        names = [d[0] for d in cur.description]
        return [{name: (str(value) if name.endswith("_id") or name in ("id", "reviewed_by") else value) for name, value in zip(names, row, strict=True)}
                for row in cur.fetchall()]


#: `list_inbox` alone joins the buyer's own account row: directive §5's "buyer identity" is what a
#: seller's inbox was missing (`GET /api/seller/requests` served `buyer_user_id`, an opaque uuid,
#: and nothing else -- the design's own fixture reads `buyer: "Dr. Rachel Mendes"`,
#: `frontend/src/logic.js:275`), and is what `list_mine`/`get_one`/`create` -- the buyer's own reads
#: two functions below -- have no business carrying about the caller themselves, so only this list
#: exists: `_COLUMNS`' own thirteen, spelled out with the `r.` qualifier the JOIN requires, plus the
#: two account columns the name is read from.
#:
#: Mirrors `app/api/admin_users.py::LIST_SQL`'s own `d.display_name`/`gb.display_name` joins -- one
#: account row, one display name, the identical "an id nobody can read" problem that entry names
#: verbatim -- with the one thing that precedent does not do for its own PRIMARY name cell
#: (`frontend/src/admin/users.ts`'s `item.name`, read straight off `a.display_name`, no fallback):
#: `request.buyer_user_id` is `NOT NULL REFERENCES account(id) ON DELETE CASCADE`
#: (`migrations/096_request.sql`), so the JOIN always finds a row, but `account.display_name`
#: itself is nullable -- a buyer who has signed up and never yet submitted an application has none
#: at all (`app/api/applications.py`'s own `COALESCE(display_name, ...)` is the only writer) -- so
#: `list_inbox` below falls back to that same account's `email`, a real identity already on file,
#: rather than showing a seller nothing, "None", or a name nobody gave.
_INBOX_COLUMNS = (
    "r.id, r.listing_id, r.buyer_user_id, r.seller_user_id, r.message, r.status,"
    " r.requested_disclosure_level, r.approved_capabilities, r.requested_at, r.reviewed_at,"
    " r.reviewed_by, r.denial_reason, r.expires_at, b.display_name, b.email"
)


def list_inbox(conn: Any, *, seller_account_id: str) -> list[dict[str, Any]]:
    with conn.cursor() as cur:
        cur.execute(
            f"SELECT {_INBOX_COLUMNS} FROM request r JOIN account b ON b.id = r.buyer_user_id"
            f" WHERE r.seller_user_id = %s ORDER BY r.requested_at DESC",
            (seller_account_id,),
        )
        names = [d[0] for d in cur.description]
        rows = []
        for values in cur.fetchall():
            row = {name: (str(value) if name.endswith("_id") or name in ("id", "reviewed_by") else value)
                   for name, value in zip(names, values, strict=True)}
            name = (row.pop("display_name") or "").strip()
            email = row.pop("email")
            row["buyer_name"] = name or email
            rows.append(row)
        return rows


def _owned_pending_or_approved(conn: Any, *, request_id: str, seller_account_id: str, required_status: tuple[str, ...]) -> None:
    """`required_status` is a TUPLE since D-C67 (2026-09-24), because `decide`'s approve arm now
    accepts an already-APPROVED row: a seller who granted three capabilities must be able to narrow
    the grant to one without withdrawing the buyer's access entirely and making them ask again.
    Every other caller passes a one-tuple and reads exactly as it did."""
    with conn.cursor() as cur:
        cur.execute("SELECT status FROM request WHERE id = %s AND seller_user_id = %s", (request_id, seller_account_id))
        found = cur.fetchone()
    if found is None:
        raise Refusal("NOT_FOUND", "No such request.", 404)  # seller_listings.py:18's rule: not yours, not found
    if found[0] not in required_status:
        wanted = " or ".join(status.lower() for status in required_status)
        raise Refusal("STATE", f"This request is {found[0].lower()}, not {wanted}.", 409)


def _chosen_capabilities(cur: Any, *, request_id: str, disclosure_level: str | None,
                        disclosure_capabilities: Sequence[str] | None) -> list[str]:
    """What this approval releases, as a sorted list of capability names — the ONE place that
    decision is made (D-C67, John, 2026-09-24).

    THE TEST ON `disclosure_capabilities` IS `is None`, NEVER TRUTHINESS, AND THAT IS THE WHOLE
    FAIL-CLOSED RULE OF THIS FAMILY. `disclosure_capabilities or covers(...)` would read an EMPTY
    list — a seller who deliberately released nothing — as "the caller said nothing", fall through
    to the requested level, and `app/api/requests.py` defaults every request to
    `FULL_CONFIDENTIAL`, so that one falsy check would silently release the practice name, the
    street, the postcode, the telephone, the exact map pin, the unredacted photographs, the
    financial packet and the floor plans together. The three arms are three different questions:

    * a SET was named (including the empty one) — that set, exactly, is what is stored;
    * a LEVEL was named — `covers()` expands it, so the existing all-five path (`FULL_CONFIDENTIAL`)
      and every single-capability one stay reachable exactly as they were;
    * neither — the level the BUYER asked for, `decide`'s own long-standing default.

    Naming both at once is refused rather than silently preferring one: two callers' worth of
    intent in one body is a caller bug, and guessing which half to honour is how a grant ends up
    wider than anybody asked for."""
    if disclosure_capabilities is not None and disclosure_level is not None:
        raise Refusal("BAD_REQUEST", "Send disclosure_capabilities or disclosure_level, not both.", 400)
    if disclosure_capabilities is not None:
        unknown = sorted({value for value in disclosure_capabilities if value not in CAPABILITIES})
        if unknown:
            raise Refusal("BAD_LEVEL", f"disclosure_capabilities may only contain {', '.join(sorted(CAPABILITIES))}.", 400)
        return sorted(set(disclosure_capabilities))
    if disclosure_level is None:
        cur.execute("SELECT requested_disclosure_level FROM request WHERE id = %s", (request_id,))
        return sorted(covers(cur.fetchone()[0]))
    if disclosure_level not in REQUESTABLE_LEVELS:
        raise Refusal("BAD_LEVEL", f"disclosure_level must be one of {', '.join(sorted(REQUESTABLE_LEVELS))}.", 400)
    return sorted(covers(disclosure_level))


def decide(conn: Any, *, request_id: str, seller_account_id: str, action: str,
          disclosure_level: str | None, reason: str | None,
          disclosure_capabilities: Sequence[str] | None = None) -> dict[str, Any]:
    """Approve or deny. Since D-C67 (John, 2026-09-24) the approve arm also accepts an ALREADY
    APPROVED row, which is how a seller narrows (or widens) what one buyer holds without
    withdrawing their access entirely; deny stays PENDING-only, because a decision the seller wants
    to take back after approving it is `revoke`, which has its own status, its own audit action and
    its own mail."""
    allowed = ("PENDING", "APPROVED") if action == "approve" else ("PENDING",)
    _owned_pending_or_approved(conn, request_id=request_id, seller_account_id=seller_account_id, required_status=allowed)
    with conn.cursor() as cur:
        if action == "approve":
            capabilities = _chosen_capabilities(
                cur, request_id=request_id, disclosure_level=disclosure_level,
                disclosure_capabilities=disclosure_capabilities)
            cur.execute(
                f"UPDATE request SET status = 'APPROVED', approved_capabilities = %s, reviewed_at = now(),"
                f" reviewed_by = %s, updated_at = now() WHERE id = %s RETURNING {_COLUMNS}",
                (capabilities, seller_account_id, request_id),
            )
        elif action == "deny":
            cur.execute(
                f"UPDATE request SET status = 'DENIED', denial_reason = %s, reviewed_at = now(),"
                f" reviewed_by = %s, updated_at = now() WHERE id = %s RETURNING {_COLUMNS}",
                (reason, seller_account_id, request_id),
            )
        else:
            raise Refusal("BAD_ACTION", "action must be 'approve' or 'deny'.", 400)
        return _row(cur)


def revoke(conn: Any, *, request_id: str, seller_account_id: str) -> dict[str, Any]:
    _owned_pending_or_approved(conn, request_id=request_id, seller_account_id=seller_account_id, required_status=("APPROVED",))
    with conn.cursor() as cur:
        cur.execute(
            f"UPDATE request SET status = 'REVOKED', updated_at = now() WHERE id = %s RETURNING {_COLUMNS}",
            (request_id,),
        )
        return _row(cur)

File: frontend/src/admin/requests.ts
/**
 * `GET /api/admin/requests` → the approved design's Requests table (Task ADMIN-REQUESTS, spec
 * `docs/superpowers/specs/2026-09-21-request-oversight-thread-design.md`, "Two prerequisites" for
 * ruling D-C63's staff oversight thread — which this module does NOT build, per that spec's own
 * scope note).
 *
 * `admin/users.ts`'s own M6 pattern, applied one tab over: a pure mapping from the API payload to
 * the rows the design's own template already renders, with `cell()` copied VERBATIM from
 * `src/logic.js`'s `adminVals()` — a FOURTH independent copy, alongside `admin/listings.ts`'s,
 * `admin/users.ts`'s and `admin/data_sources.ts`'s own, so a drift on any one side fails on its
 * own rather than one copy silently covering for another.
 *
 * **This tab has NO Action column at all.** The design's own `columns` for the "activity" tab are
 * `["Request", "Practice", "Status", "Age"]` — Users and Listings both carry a fifth, `"Decision"`/
 * `"Action"`, and this one does not, because staff does not DECIDE a disclosure request: the
 * seller does, through `app/api/seller_requests.py::decide_request`. So `toRequestRows` below
 * produces four `Cell`s per row and none of them ever carries an action, and there is no
 * reload-after-decision seam for this module to wire the way `admin/users.ts`'s `list(reload)`
 * does — `AdminRequestsAdapter.list()` takes no argument.
 *
 * **What this module deliberately does not render.** The design's own footnote on this tab
 * (`sets.activity.footnote`) says "Message contents are visible only in an abuse investigation,
 * and every such view is logged" — `app/api/admin_requests.py` does not serve `request.message` at
 * all, for exactly that reason, so there is nothing here to withhold a second time. The Request
 * column's sub-line instead names the disclosure LEVEL asked for (or, once a seller has decided,
 * the level actually granted) — real data the route does serve, and the one fact this column can
 * state without opening a door the design's own prose says is investigation-only.
 */
import { LEVEL_LABEL, orderedCapabilities } from '../disclosure/capabilities';

export interface Cell {
  hasMain: boolean; main: string;
  hasSub: boolean; sub: string;
  hasPill: boolean; pill: string; pillStyle: string;
  hasActions: boolean; actions: never[];
}

/** Verbatim from logic.js's `adminVals()`, narrowed to the shape this tab ever uses: `actions` is
 *  always empty here (module note), so the fourth `cell()` parameter is never passed and its
 *  `never[]` type documents that rather than merely defaulting it. */
export function cell(main: string | null, sub?: string | null, pill?: string | null, pillTone?: string | null): Cell {
  const tones: Record<string, string[]> = {
    ok: ['#ffffff', '#003a70', '#003a70'], warn: ['#003a70', '#deecf7', '#deecf7'],
    bad: ['#494949', '#ffffff', '#494949'], info: ['#003a70', '#ffffff', '#339dde'],
    mute: ['#494949', '#f5f5f5', '#d4dde5']
  };
  const t = tones[pillTone ?? ''] || tones.mute;
  return {
    hasMain: !!main, main: main || '', hasSub: !!sub, sub: sub || '',
    hasPill: !!pill, pill: pill || '',
    pillStyle: 'display: inline-block; font-size: 11.5px; font-weight: 500; padding: 4px 11px; border-radius: 999px; color: ' + t[0] + '; background: ' + t[1] + '; border: 1px solid ' + t[2] + ';',
    hasActions: false, actions: []
  };
}


/** `request.status`'s own CHECK constraint (`migrations/096_request.sql`), pinned by equality in
 *  `tests/test_docs.py` against `app/api/admin_requests.py::STATUSES`. `[label, tone]` per value:
 *  PENDING/APPROVED/DENIED take the design's OWN three words and tones off its "activity" fixture
 *  (rows 1-3: "Awaiting seller" warn, "Engaged" ok, "Declined" bad); REVOKED has no design fixture
 *  to read a word from — A53's own precedent one surface over (the seller inbox pill) gives a
 *  revoked row its OWN word, "Revoked", never collapsed onto "Declined" — and takes the same `bad`
 *  tone as Declined, since neither means the buyer is currently disclosed to. */
export const STATUS_PILL: Record<string, [string, string]> = {
  PENDING: ['Awaiting seller', 'warn'],
  APPROVED: ['Engaged', 'ok'],
  DENIED: ['Declined', 'bad'],
  REVOKED: ['Revoked', 'bad']
};

/** MOVED to `frontend/src/disclosure/capabilities.ts` under ruling D-C67 (2026-09-24), which gave
 *  these words a second reader — the seller's own per-capability chooser — and re-exported here so
 *  every existing importer of this module reads the same one table rather than a second copy of
 *  it. */
export { LEVEL_LABEL } from '../disclosure/capabilities';

const MONTHS = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

/** "September 3" from an ISO timestamp — `admin/users.ts::formatDate`'s own shape, UTC and never
 *  `Date#toLocaleDateString`, for the reason that module records: the runtime's own locale must
 *  not move the day the API actually stamped. */
function formatDate(iso: string): string {
  const d = new Date(iso);
  return `${MONTHS[d.getUTCMonth()]} ${d.getUTCDate()}`;
}

const DAY_MS = 24 * 60 * 60 * 1000;

/** Whole days between `iso` and now, floored — never fractional ("6.3 days"), and never negative
 *  (a clock skew of a few seconds between this browser and the server that stamped `requested_at`
 *  must not read "-1 days"). */
function ageDays(iso: string): number {
  return Math.max(0, Math.floor((Date.now() - new Date(iso).getTime()) / DAY_MS));
}

function plural(n: number, noun: string): string {
  return `${n} ${noun}${n === 1 ? '' : 's'}`;
}

/**
 * One row of `GET /api/admin/requests`'s `items`, narrowed to what the table reads. A structural
 * interface, so the real payload is assignable without restating the fields this tab does not
 * print (`app/api/admin_requests.py` serves more: `listing_id`, `buyer_user_id`, `seller_user_id`,
 * kept here anyway because a caller may need them even though no cell reads them directly).
 */
export interface AdminRequestItem {
  id: string;
  status: string;
  requested_disclosure_level: string;
  /** The SET the seller released (`request.approved_capabilities`, migration 097) — one name, five
   *  names, or the empty array a seller who approved and released nothing leaves behind. `null`
   *  while no decision has been taken, or where the decision was a refusal. */
  approved_capabilities: string[] | null;
  requested_at: string;
  reviewed_at: string | null;
  listing_id: string;
  listing_name: string | null;
  listing_type: string | null;
  listing_city: string | null;
  buyer_user_id: string;
  buyer_name: string;
  seller_user_id: string;
  seller_name: string;
}

/**
 * One of the DESIGN's own `sets.activity.rows` fixtures, as `frontend/tests/design-admin-requests.mjs`
 * serves them — `admin/users.ts`'s `DesignUserRow` applied to `admin-requests`, one of the thirteen
 * frozen screens. The fixture carries every one of the design's own WORDS AND STYLES directly, read
 * straight off `adminVals()`'s own computed cells — `pillStyle` included — rather than a tone NAME
 * this module would have to re-derive: the `.mjs` fixture files touch only `logic.js`, never a
 * TypeScript module, and a style string copied verbatim off the design's own computed output can
 * never drift from it.
 */
export interface DesignRequestRow {
  id: string;
  request: string; requestSub: string;
  practice: string; practiceSub: string;
  pill: string; pillStyle: string;
  age: string; ageSub: string;
}

/** The badge: every request still `PENDING` — the one status still waiting on a seller to act —
 *  over the WHOLE table, as `app/api/admin_requests.py::queue_counts` serves it beside `items`. */
export interface AdminRequestCounts { pending: number; total: number }

interface QueuePage { items?: unknown; next_cursor?: string | null; counts?: unknown }

/** The counts, or null where the answer carried none this table can read — an older API, a shape
 *  nobody has seen, or the explicit `counts: null` the route serves on every page past the first
 *  (the count describes the whole table, so a paging caller already holds it). `admin/users.ts`'s
 *  own `countsOf` shape and its own reason: `!= null`, not `!== undefined`, because a JSON `null`
 *  passes an undefined test and `null.pending` throws inside the loader. */
export function countsOf(body: QueuePage): AdminRequestCounts | null {
  const counts = body.counts as AdminRequestCounts | null | undefined;
  return counts != null && typeof counts.pending === 'number' && typeof counts.total === 'number' ? counts : null;
}

/** Given/approved disclosure level, or requested — the module note's own choice for what the
 *  Request column's sub-line names, since the raw message is never served at all. */
function levelLine(item: AdminRequestItem): string {
  // `!= null`, never truthiness: an APPROVED row whose seller released NOTHING carries `[]`, which
  // is a real decision and must read as one rather than falling back to what the buyer asked for
  // (D-C67's own fail-closed rule, on the surface that merely reports it).
  const approved = item.status === 'APPROVED' && item.approved_capabilities != null;
  if (approved) {
    // A LIST, not the seller inbox's SENTENCE: this is a table cell, so it keeps this tab's own
    // Title-case register (`LEVEL_LABEL`) and joins with commas, where `capabilityPhrase` is
    // lower-case and ends in "and" because it is read inside a sentence one surface over.
    // No fallback: `orderedCapabilities` answers only names `LEVEL_LABEL` has a word for. The
    // `?? level` on the REQUESTED line below stays, because that column carries whatever the API
    // serves and an unrecognised level there is a real possibility rather than a dead branch.
    const names = orderedCapabilities(item.approved_capabilities).map((name) => LEVEL_LABEL[name]);
    return names.length === 0 ? 'Approved: nothing released' : `Approved: ${names.join(', ')}`;
  }
  const level = item.requested_disclosure_level;
  return `Requested: ${LEVEL_LABEL[level] ?? level}`;
}

/** The practice's own name, or `admin/listings.ts`'s own `type practice — city` composed label, or
 *  its own "Untitled listing" fallback — the identical idiom, applied to this tab's own fields. */
function practiceLabel(item: AdminRequestItem): string {
  const label = item.listing_city && item.listing_type ? `${item.listing_type} practice — ${item.listing_city}` : null;
  return item.listing_name || label || 'Untitled listing';
}

/** The Age column: whole days since the request, and — only for a request a seller has actually
 *  decided — when. No fabricated narrative ("Reminder sent", "Packet released"): the design's own
 *  four fixture rows carry one, but no real feature backs any of the three phrases the design
 *  invented for it, and D-C53's "zero-fake-data" rule is what this column follows instead. */
function ageCell(item: AdminRequestItem): Cell {
  const days = plural(ageDays(item.requested_at), 'day');
  return cell(days, item.reviewed_at === null ? null : `Reviewed ${formatDate(item.reviewed_at)}`);
}

/**
 * Exactly the shape of `sets.<tab>.rows` in `adminVals()` — an array of cell arrays
 * (`admin/users.ts`'s own Important-3 note: `set.rows.map((cells, i) => ({ cells, style }))` is
 * what wraps them with the grid, so returning anything else would lose `style` and re-flow the
 * table at `maxDiffPixels: 0`). `admin-requests` IS one of the thirteen frozen screens, so this
 * function IS reached by the pixel oracle — through the `DesignRequestRow` arm, which is where
 * every one of its captured pixels comes from.
 */
export function toRequestRows(items: (AdminRequestItem | DesignRequestRow)[]): Cell[][] {
  return items.map((item) => {
    if ('request' in item) {
      return [
        cell(item.request, item.requestSub),
        cell(item.practice, item.practiceSub),
        { ...cell(null, null, item.pill), pillStyle: item.pillStyle },
        cell(item.age, item.ageSub)
      ];
    }
    const [pill, tone] = STATUS_PILL[item.status] ?? [item.status, 'mute'];
    return [
      cell(item.buyer_name, levelLine(item)),
      cell(practiceLabel(item), item.seller_name),
      cell(null, null, pill, tone),
      ageCell(item)
    ];
  });
}

/** One page of the queue, and far past what any real one holds — `admin/users.ts`'s own
 *  `PAGE_LIMIT`/`MAX_PAGES` shape and reason: a server that answers with the cursor it was given
 *  would otherwise spin for ever. 200 is `admin_requests.MAX_LIST`, so a page is never short of
 *  what was asked for. */
const PAGE_LIMIT = 200;
const MAX_PAGES = 20;

/** What `logic.js` sees as `this.props.adminRequests`. No `reload` argument, unlike
 *  `AdminUsersAdapter.list(reload)`: this tab has no per-row decision that could need one
 *  (module note). */
export interface AdminRequestsAdapter {
  list(): Promise<{ rows: Cell[][]; counts: AdminRequestCounts | null }>;
}

export function makeAdminRequestsAdapter(): AdminRequestsAdapter {
  return {
    list: async () => {
      // NO `X-CSRF-Token` and NO `Content-Type`: this is a read — `admin/data_sources.ts`'s own
      // rule (`check_origin_and_csrf` returns early on GET/HEAD/OPTIONS, so both would be refused
      // on an empty value in any case).
      const items: (AdminRequestItem | DesignRequestRow)[] = [];
      let counts: AdminRequestCounts | null = null;
      let cursor: string | null = null;
      for (let page = 0; page < MAX_PAGES; page++) {
        const query = `/api/admin/requests?limit=${PAGE_LIMIT}${cursor === null ? '' : `&cursor=${encodeURIComponent(cursor)}`}`;
        const res = await fetch(query, { method: 'GET', credentials: 'same-origin' });
        if (!res.ok) throw new Error('the request queue could not be read');
        const body = (await res.json()) as QueuePage;
        if (!Array.isArray(body.items)) throw new Error('the request queue answered no items');
        items.push(...(body.items as (AdminRequestItem | DesignRequestRow)[]));
        if (counts === null) counts = countsOf(body);
        cursor = body.next_cursor ?? null;
        if (cursor === null) break;
      }
      return { rows: toRequestRows(items), counts };
    }
  };
}
